add method log_lbfgs - Riemannian logarithm using expm_frechet -

dnguyend opened the pull request:
I added log_lbfgs - please let me know! I added new tests to  [test/manifolds/stiefel.jl](https://github.com/dnguyend/Manifolds.jl/blob/master/test/manifolds/stiefel.jl). Additional tests are in  [this workbook](https://github.com/dnguyend/MiscCollection/blob/main/MiscStiefelLogTest/misc_stiefel_frechet/JuliaStiefelLogMiminizeLBFGS.ipynb) and the jl files in [this directory](https://github.com/dnguyend/MiscCollection/tree/main/MiscStiefelLogTest/misc_stiefel_frechet)

kellertuer commented:
The first thing that is included, is, that your files have 0% code coverage, that you will have to check.

Concerning the format, the CI reports a lot of changes necessary:
https://github.com/JuliaManifolds/Manifolds.jl/actions/runs/3578455917/jobs/6018629102

But to fix that make sure
(a) to have Juliaformatter in its newest version
(b) to run the formatting from the base folder of Manifolds, e.g. as `using JuliaFormatter; format(".")`, because we have a few setups for our formatting locally as well.

Besides that – wow! 600 lines of code for one function.

dnguyend commented:
Thanks, I will give it another try - I see there is zero coverage for other
manifolds which I did not touch - eg SymmetricPositiveDefiniteFixedRank and
I am not sure where to start looking. The only test script I changed was
stiefel.jl. So I do need some help here. I reran the formatter.

The function itself is around 200 lines of code, including helpers (eg
vectorize/unvectorize of antisymmetric matrices). The 600 lines of code is
because I implemented a few functions that I believe should belong to
common Julia libraries - but I do not see them anywhere. They should be
useful outside of Manifolds.jl, and presumably could be included elsewhere.
* expm_frechet
* minimize_lbfgs (this is in Optim, but we talked about Optim being
expensive to include)
* I also included dot_exp, which gives us the velocity of the geodesic, and
included an alternative formula for the geodesic.


On Wed, Nov 30, 2022 at 7:29 AM Ronny Bergmann ***@***.***>
wrote:

> The first thing that is included, is, that your files have 0% code
> coverage, that you will have to check.
>
> Concerning the format, the CI reports a lot of changes necessary:
>
> https://github.com/JuliaManifolds/Manifolds.jl/actions/runs/3578455917/jobs/6018629102
>
> But to fix that make sure
> (a) to have Juliaformatter in its newest version
> (b) to run the formatting from the base folder of Manifolds, e.g. as using
> JuliaFormatter; format("."), because we have a few setups for our
> formatting locally as well.
>
> Besides that – wow! 600 lines of code for one function.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332073265>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFP2KPQ4ELPRSLGSAGTWK5CD5ANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
The other manifolds will work as soon as formatting is fine. As soon as a first of the tests fails, no other is started and since we have groups of tests, some were not run (see those with a gray (!) upfront in the list of the checks, those were not run) so as soon as those do run, the existing manifolds are covered (well 98.94% of their code).

Those new functions I will have to check, especially `dot_exp` is already provided I think.

dnguyend commented:
Thanks - I will submit the reformatted files - please bear with me if there
are further failed mails. If there is a quieter way for me to test before
resubmitting please let me know.
Du

On Wed, Nov 30, 2022 at 8:28 AM Ronny Bergmann ***@***.***>
wrote:

> The other manifolds will work as soon as formatting is fine. As soon as a
> first of the tests fails, no other is started and since we have groups of
> tests, some were not run (see those with a gray (!) upfront in the list of
> the checks, those were not run) so as soon as those do run, the existing
> manifolds are covered (well 98.94% of their code).
>
> Those new functions I will have to check, especially dot_exp is already
> provided I think.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332153369>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFPO27USH55INOFOCY3WK5I6BANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
Well if you have Manifolds on you machine in development mode, you can always also test your code locally with `] test Manifolds`. But even if you only let the CI here check – that is still reasonably silent ;)

dnguyend commented:
Thanks for the tip!

On Wed, Nov 30, 2022 at 8:56 AM Ronny Bergmann ***@***.***>
wrote:

> Well if you have Manifolds on you machine in development mode, you can
> always also test your code locally with ] test Manifolds. But even if you
> only let the CI here check – that is still reasonably silent ;)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332188483>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFNI2QAVMSXHNK6SYLTWK5MIZANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
Beside the tests in stiefel.jl, I did extensive tests in this repository
<https://github.com/dnguyend/MiscCollection/tree/main/MiscStiefelLogTest/misc_stiefel_frechet>
,
in particular, this workbook
<https://github.com/dnguyend/MiscCollection/blob/main/MiscStiefelLogTest/misc_stiefel_frechet/JuliaStiefelLogMiminizeLBFGS.ipynb>



On Wed, Nov 30, 2022 at 12:04 PM Du Nguyen ***@***.***> wrote:

> I did a push - it seems OK - please let me know!
>
> On Wed, Nov 30, 2022 at 8:57 AM Du Nguyen ***@***.***> wrote:
>
>> Thanks for the tip!
>>
>> On Wed, Nov 30, 2022 at 8:56 AM Ronny Bergmann ***@***.***>
>> wrote:
>>
>>> Well if you have Manifolds on you machine in development mode, you can
>>> always also test your code locally with ] test Manifolds. But even if
>>> you only let the CI here check – that is still reasonably silent ;)
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332188483>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AEOOGFNI2QAVMSXHNK6SYLTWK5MIZANCNFSM6AAAAAASO7CVTA>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>
>>
>> --
>> Thanks,
>> Du
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

kellertuer commented:
Formatting seems fine (you do not have to write a comment for every commit).

Concerning the tests, well here in the CI we aim to have an – maybe small in dimension but – test that covers every line and still performs something we can check to be valid/correct. This doe snot say we do not have bugs, but (a) minimises the number of bugs, since you have to think about code lines a second time, namely what you expect as an outcome an write a test for that, but even more
(b) the automatic tests (with high test coverage) make It easier to fix bugs and modify/improve the code, because this way we can be sure it still “behaves as expected”.

dnguyend commented:
Ok, thanks! I hope the tests added in stiefel.jl conform to the general
philosophy. Happy to make any change if required.
I am quite impressed with the framework you guys set up.

On Wed, Nov 30, 2022 at 12:31 PM Ronny Bergmann ***@***.***>
wrote:

> Formatting seems fine (you do not have to write a comment for every
> commit).
>
> Concerning the tests, well here in the CI we aim to have an – maybe small
> in dimension but – test that covers every line and still performs something
> we can check to be valid/correct. This doe snot say we do not have bugs,
> but (a) minimises the number of bugs, since you have to think about code
> lines a second time, namely what you expect as an outcome an write a test
> for that, but even more
> (b) the automatic tests (with high test coverage) make It easier to fix
> bugs and modify/improve the code, because this way we can be sure it still
> “behaves as expected”.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332509002>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFN5PGINA6SLBKMREVDWK6FPTANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
The CI seems to have failed on installing matplotlib - which I don't think
I use. What should I do?

On Wed, Nov 30, 2022 at 12:49 PM Du Nguyen ***@***.***> wrote:

> Ok, thanks! I hope the tests added in stiefel.jl conform to the general
> philosophy. Happy to make any change if required.
> I am quite impressed with the framework you guys set up.
>
> On Wed, Nov 30, 2022 at 12:31 PM Ronny Bergmann ***@***.***>
> wrote:
>
>> Formatting seems fine (you do not have to write a comment for every
>> commit).
>>
>> Concerning the tests, well here in the CI we aim to have an – maybe small
>> in dimension but – test that covers every line and still performs something
>> we can check to be valid/correct. This doe snot say we do not have bugs,
>> but (a) minimises the number of bugs, since you have to think about code
>> lines a second time, namely what you expect as an outcome an write a test
>> for that, but even more
>> (b) the automatic tests (with high test coverage) make It easier to fix
>> bugs and modify/improve the code, because this way we can be sure it still
>> “behaves as expected”.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1332509002>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEOOGFN5PGINA6SLBKMREVDWK6FPTANCNFSM6AAAAAASO7CVTA>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

kellertuer commented:
Work further locally on tests and documentation until

https://github.com/JuliaPy/Conda.jl/issues/229

is dressed / fixed :)

dnguyend commented:
Thanks, I saw the rerun, and I think there is one still going on.
On a rerun there was a fail comparing
log_lbfgs vs generating vector for the canonical metric - they look
sufficiently close (less than 9e-9) but failed - I believe in the local
test I forced the accuracy level to be higher - I probably will do it here.

There was an error for expm_frechet - absolute error was < 3e-3, probably
on a matrix with large eigenvalue. I will limit the largest eigenvalue and
resubmit the test script.

There was also a complaint on symplectic stiefel - but I presume that is
not related.
Have a good weekend!


α=0.5: Test Failed at
/home/runner/work/Manifolds.jl/Manifolds.jl/test/manifolds/stiefel.jl:568
  Expression: isapprox(MM, p, Manifolds.log_lbfgs(MM, p, q), X)
   Evaluated: isapprox(MetricManifold(Stiefel(4, 3, ℝ),
StiefelSubmersionMetric{Float64}(0.5)), [0.03692676533179337
0.7392015040180194 -0.6496424690154912; 0.38776641156841574
-0.24069024074501766 -0.013910854515959737; 0.8809850680488652
0.25081693802237826 0.2519849355176593; 0.26858692063542816
-0.5768363789400607 -0.7171295160919356], [0.0585973393258645
0.08436281749191825 0.11905892082509396; -0.12976266615385246
-0.6930079176321513 -0.7317602195557313; 0.15076133210137377
-0.13558784475699176 0.07866681264844064; -0.31522268101696455
0.3383171436064817 -0.06601801843791485], [0.05859734123700349
0.08436281763177483 0.1190589231066079; -0.12976266057166255
-0.693007929702738 -0.7317602231596487; 0.15076133133869893
-0.13558784471759106 0.07866681181496862; -0.31522268683725074
0.3383171488393988 -0.06601802072767841])

On Thu, Dec 1, 2022 at 12:34 AM Ronny Bergmann ***@***.***>
wrote:

> Work further locally on tests and documentation until
>
> JuliaPy/Conda.jl#229 <https://github.com/JuliaPy/Conda.jl/issues/229>
>
> is dressed / fixed :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1333219690>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFJHN7SFLU7247GFTTTWLA2FTANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
Well your tests do not yet have reported their coverage, we have to wait for that still. But none of the tests should be failing. Please check all fails (our previous ones should be very stable and fine).

dnguyend commented:
Yes - will push with the changes indicated in the previous email later
today.
Thanks
Du

On Fri, Dec 2, 2022 at 9:22 AM Ronny Bergmann ***@***.***>
wrote:

> Well your tests do not yet have reported their coverage, we have to wait
> for that still. But none of the tests should be failing. Please check all
> fails (our previous ones should be very stable and fine).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1335305280>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFKLCTOYZV7E2YATIBDWLIA3LANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
I just pushed - I adjusted atol to 1e-6 on the comparison and loop the test
1000 times locally - seems OK - I hope it will go through in the CI run.
Please bear with me if there are further issues.

On Fri, Dec 2, 2022 at 9:40 AM Du Nguyen ***@***.***> wrote:

> Yes - will push with the changes indicated in the previous email later
> today.
> Thanks
> Du
>
> On Fri, Dec 2, 2022 at 9:22 AM Ronny Bergmann ***@***.***>
> wrote:
>
>> Well your tests do not yet have reported their coverage, we have to wait
>> for that still. But none of the tests should be failing. Please check all
>> fails (our previous ones should be very stable and fine).
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1335305280>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEOOGFKLCTOYZV7E2YATIBDWLIA3LANCNFSM6AAAAAASO7CVTA>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

dnguyend commented:
The CI run seems OK - the nightly failed on symplecticstiefel, which is a
different manifold that I did not make any change. I think this does not
relate to log_lbfgs. It is likely the setup for the nightly is different
from the CI run. Please let me know if there is anything I should change.
Thanks,
Du

/home/runner/work/Manifolds.jl/Manifolds.jl/test/manifolds/symplecticstiefel.jl:222
  Test threw exception
  Expression: is_vector(SpSt_6_4, p_6_4, A_6_4_proj, true; atol = 1.0e-12)
  DomainError with 1.0406995415442994e-12:
  The matrix X is not in the tangent space at point p of the
SymplecticStiefel{6, 4, ℝ}() manifold, as p^{+}X is not a Hamiltonian
matrix.

On Fri, Dec 2, 2022 at 10:39 AM Du Nguyen ***@***.***> wrote:

> I just pushed - I adjusted atol to 1e-6 on the comparison and loop the
> test 1000 times locally - seems OK - I hope it will go through in the CI
> run. Please bear with me if there are further issues.
>
> On Fri, Dec 2, 2022 at 9:40 AM Du Nguyen ***@***.***> wrote:
>
>> Yes - will push with the changes indicated in the previous email later
>> today.
>> Thanks
>> Du
>>
>> On Fri, Dec 2, 2022 at 9:22 AM Ronny Bergmann ***@***.***>
>> wrote:
>>
>>> Well your tests do not yet have reported their coverage, we have to wait
>>> for that still. But none of the tests should be failing. Please check all
>>> fails (our previous ones should be very stable and fine).
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1335305280>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AEOOGFKLCTOYZV7E2YATIBDWLIA3LANCNFSM6AAAAAASO7CVTA>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>
>>
>> --
>> Thanks,
>> Du
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

kellertuer commented:
Nightly might fail, that is ok, we have that one to “look ahead” a little. If here on Github you look at the 2 codecov tests, you currently have a 81.6% coverage for your new code (but you need 98.94) the same for the overall code average, where you are missing 1.15% (but getting the patch one to 100% fixes that.

You can see which lines you are missing at https://app.codecov.io/gh/JuliaManifolds/Manifolds.jl/pull/555/diff - for example minimize_lbfgs Curren try has just around 80% coverage (the `@view`s are false positives but there are a lot of special cases uncovered in the algorithms).

I will already turn on documentation rendering here, because all functions need to be documented and be included in the docs.

kellertuer commented on src/expm_frechet.jl:
```diff
@@ -0,0 +1,532 @@
+
+@doc raw"""
+based on
+.. [1] Awad H. Al-Mohy and Nicholas J. Higham (2009)
```
Please check how we do references in footnotes.

dnguyend replied:
This is fixed - I have generated the doc locally - but couldnt find the doc link generated by CI
Manifolds.expm_frechet
— Function

expm_frechet(A, E)
expm_frechet!(buff, A, E)

Compute Frechet derivative of expm(A) in direction E using algorithm 6.4 of [AlMohyHigham2009]

expm_frechet(A,E)=lim⁡t→01t(exp⁡(A+tE)−exp⁡(A))expm\_frechet(A, E) = \lim_{t\to 0}\frac{1}{t}(\exp(A + tE) - \exp(A))
expm_frechet(A,E)=t→0lim​t1​(exp(A+tE)−exp(A))

For sufficiently small ∣A∣1|A|_1∣A∣1​ norm, we use Padé with an appropriate number of terms ( one of (3, 5, 7, 9, 13)), with 13 terms is the default. Otherwise, we use the formula exp(A) = (exp(A/2^s))^{2^s}, where s is used to scale so 2−s∣A∣12^{-s}|A|_12−s∣A∣1​ is smaller than ell_table_61ell\_table\_61ell_table_61[14] of the lookup table ell_table_61ell\_table\_61ell_table_61 in the code, which comes from table 6.1 in op. cit.

For expm_frechet!expm\_frechet!expm_frechet!, buff is a matrix of size 16*k times k, where kkk is the size of AAA. The returns, (exp(A), dexp(A, E)) are stored in the first two blocks of buff. The remaining blocks are used as temporary storage. This is for algorithms calling expm_frechet repeatedly, where allocation becomes the biggest bottleneck.

kellertuer replied:
please use (with `@docs raw"""`to start the string for the string Math formulae as actual math – in markdown that is two backtracks, so something line ``` ``\lVert A \rVert_1`` ``` for the matrix 1-norm in the doc string

dnguyend replied:
Thanks - fixed - actually the cut and paste (here and below) somehow corrupted the HTML format - it should looks better on the docs webpage.

kellertuer commented on src/expm_frechet.jl:
```diff
@@ -0,0 +1,532 @@
+           30 (4). pp. 1639-1657. ISSN 1095-7162
+"""
+
+ell_table_61 = (
```
what are these magic numbers? They do not even have a comment what they are for?

dnguyend replied:
Added comments - there are some typos - will change.

# From table 6.1 of [^AlMohyHigham2009]. A list of cut off values $l_m$ used to determine how many terms of Padé approximant is used to compute expm_frechet(A, E), look for the largest i\
   ndex $m\in [3, 5, 7, 9, 13]$ just exceed $|A|_1$.
ell_table_61 = (

kellertuer replied:
The comment in the code contains a few typos, but I am really not happy with such magic numbers somewhere in the code. And especially a Nothing upfront and such.

The problem is, that we as maintainers have to be able to take care of this code. And even in the functions below there is quite a few “magic numbers”. I am really not happy about this.

dnguyend replied:
Instead of the table I can write a function implementing the cutoff logic of how many pade terms to use. Will have to think about this. It may not be straightforward, otherwise it is already done in that paper.

Pade coefficients can be generated from a function.  I can construct a expm_pade object, and we initialize it with coefficients then pass to functions that call expm_frechet - that way we dont have to save the coefficients, and dont have to keep regenerating them - it may be a bit more complicated to use the function, though. Appreciate if you have a suggestion.

kellertuer commented on src/expm_frechet.jl:
```diff
@@ -0,0 +1,532 @@
+@doc raw"""
+    _diff_pade3(A, E)
+
+Compute expm using 3-term pade approximant
```
hm, sorry, this might be too less information, I am not sure what this function does.

dnguyend replied:
expanded docstring
@doc raw"""
 34    _diff_pade3(A, E)
 35    Compute U, V, LU, LV from the first 6 lines of the for loop in
 36    algorithm 6.4 of [^AlMohyHigham2009] using 3-term Padé approximant
 37    the tuple b contains the Padé coefficients of the numerator
 38"""
 39@inline function _diff_pade3!(buff, A, E)
 40    b = (120.0, 60.0, 12.0, 1.0)

kellertuer replied:
line 34 needs to be indented by four letters and followed by an empty line to render correctly.

dnguyend replied:
thanks - this is fixed - BTW, this function is internal, it does not show on the doc. - see more below

kellertuer replied:
Internal or not, as a maintainer I still have to understand what it does and avoid that it has bugs, because the issues here in this repository are things _I_ have to answer and fix. So for that internal or external is not an issue, stability and clarity (and bug-free functions ;)) are the main issue.

kellertuer commented on src/minimize_lbfgs.jl:
```diff
@@ -0,0 +1,506 @@
+using LinearAlgebra
```
Please collect using-statements in `Manifolds.jl` main file (I think we already have LinearAlgebra in there) – and for best of cases state the functions you need explicitly, to keep the “inner name space” of Manifolds.jl not too crowded.

dnguyend replied:
removed

kellertuer commented on src/manifolds/StiefelFrechet.jl:
```diff
@@ -0,0 +1,312 @@
+        return copyto!(mat21, R)
+    end
+
+    function fun!(F, G, v)
```
Can you look for a better function name here and add documentation as well?

kellertuer commented on src/manifolds/StiefelFrechet.jl:
```diff
@@ -0,0 +1,312 @@
+_g_rank_cutoff = 1e-14  # cut off of eigenvalue to determine rank of span of two orthogonal basis
```
These first seven lines contain some “magic numbers” that not only have names I do not understand, but even more it would be better if a user could set these. Can you put them into their corresponding functions as keyword arguments please?

dnguyend replied:
they actually could be overwritten in lbfgs_options - I changed the codes to make it clearer. In the docstring I explained what the variable means 

102    if !isnothing(lbfgs_options)
103        complementary_rank_cutoff = get(lbfgs_options, "complementary_rank_cutoff", 1e-14)
104        corrections = get(lbfgs_options, "corrections", 10)
105        c1 = get(lbfgs_options, "c1", 1e-4)
106        c2 = get(lbfgs_options, "c2", 0.9)
107        max_ls = get(lbfgs_options, "max_ls", 25)
108        max_fun_evals = get(lbfgs_options, "max_fun_evals", Int(ceil(max_itr * 1.3)))
109    else
110        complementary_rank_cutoff = 1e-14
111        corrections = 10
112        c1 = 1e-4
113        c2 = 0.9
114        max_ls = 25
115        max_fun_evals = Int(ceil(max_itr * 1.3))
116    end

From the doc string (sorry the HTML paste doesnt work well here - but should be visible on the website).
lbfgs_options is a dictionary of technical parameters, isolated to a separate group as we do not expect users to modify them. The relevant keys are

    complementary_rank_cutoffcomplementary\_rank\_cutoffcomplementary_rank_cutoff (default = 1e-14): cut off of eigenvalue to determine the rank of the complementary basis QQQ. We use SVD to determine the size of the complement basis QQQ of YYY in span(Y,W)span(Y, W)span(Y,W). We keep columns of QQQ corresponding to eigenvalues higher than this cutoff.

    The remaining parameters are related to the L-BFGS minimizer minimize:
        correctionscorrectionscorrections (default = 10): default number of stored history in L-BFGS.
        c1c1c1 (default = 1e-4) default c1, c2 in the line search.
        c2c2c2 (default = 0.9).
        max_lsmax\_lsmax_ls (default = 25) max iterations for the line search.
        maxfunevalsmax_fun_evalsmaxf​une​vals (default = Int(ceil(max_itr * 1.3))) max function evaluations

kellertuer reviewed (commented):
I just saw a few lines when looking for the Codecoverage Link – mainly I noticed the ones where I mentioned “magic numbers”. I did not yet have time to read all your code thoroughly, but maybe these first comments give an impression what we roughly aim for in code comments and usability.

dnguyend commented:
Thanks for pointing that out - I will make changes - is there a quick
command to preview the doc ? I am on emacs and not any IDE, btw. I will
search on google otherwise.

Just quickly, if its relevant ( will add to the doc), the magic numbers on
expm_frechet comes from pade approximation (copied from scipy.linalg), the
idea of the algorithm from Higham is expm_frechet is done similar to expm,
we do pade for small norm matrices then us the identity exp(x) = exp(x/2)^2
for larger norm.

On Fri, Dec 2, 2022 at 3:32 PM Ronny Bergmann ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> I just saw a few lines when looking for the Codecoverage Link – mainly I
> noticed the ones where I mentioned “magic numbers”. I did not yet have time
> to read all your code thoroughly, but maybe these first comments give an
> impression what we roughly aim for in code comments and usability.
> ------------------------------
>
> In src/expm_frechet.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038506284>
> :
>
> > @@ -0,0 +1,532 @@
>
> +
>
> ***@***.*** raw"""
>
> +based on
>
> +.. [1] Awad H. Al-Mohy and Nicholas J. Higham (2009)
>
>
> Please check how we do references in footnotes.
> ------------------------------
>
> In src/expm_frechet.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038506496>
> :
>
> > @@ -0,0 +1,532 @@
>
> +
>
> ***@***.*** raw"""
>
> +based on
>
> +.. [1] Awad H. Al-Mohy and Nicholas J. Higham (2009)
>
> +           Computing the Frechet Derivative of the Matrix Exponential,
>
> +           with an application to Condition Number Estimation.
>
> +           SIAM Journal On Matrix Analysis and Applications.,
>
> +           30 (4). pp. 1639-1657. ISSN 1095-7162
>
> +"""
>
> +
>
> +ell_table_61 = (
>
>
> what are these magic numbers? They do not even have a comment what they
> are for?
> ------------------------------
>
> In src/expm_frechet.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038506806>
> :
>
> > +    3.13e0,
>
> +    3.90e0,
>
> +    4.74e0,
>
> +    5.63e0,
>
> +    6.56e0,
>
> +    7.52e0,
>
> +    8.53e0,
>
> +    9.56e0,
>
> +    1.06e1,
>
> +    1.17e1,
>
> +)
>
> +
>
> ***@***.*** raw"""
>
> +    _diff_pade3(A, E)
>
> +
>
> +Compute expm using 3-term pade approximant
>
>
> hm, sorry, this might be too less information, I am not sure what this
> function does.
> ------------------------------
>
> In src/minimize_lbfgs.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038507628>
> :
>
> > @@ -0,0 +1,506 @@
>
> +using LinearAlgebra
>
>
> Please collect using-statements in Manifolds.jl main file (I think we
> already have LinearAlgebra in there) – and for best of cases state the
> functions you need explicitly, to keep the “inner name space” of
> Manifolds.jl not too crowded.
> ------------------------------
>
> In src/manifolds/StiefelFrechet.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038508258>
> :
>
> > +
>
> +    # unravel v to A and R
>
> +    @inline function unvecAR!(A, R, v)
>
> +        A .= unveca(v[1:Adim])
>
> +        return R .= reshape(v[(Adim + 1):end], k, p)
>
> +    end
>
> +
>
> +    # compute ex1= exp((1-2*alf)*A) and populate
>
> +    # the (p+k) × (p+k) matrix mat in the Exp formula. ex2=exp(mat)
>
> +    @inline function setMatForExp!(mat, ex1, ex2, A, R)
>
> +        mul!(mat11, 2 * alf, A)
>
> +        mul!(mat12, -1, R')
>
> +        return copyto!(mat21, R)
>
> +    end
>
> +
>
> +    function fun!(F, G, v)
>
>
> Can you look for a better function name here and add documentation as well?
> ------------------------------
>
> In src/manifolds/StiefelFrechet.jl
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#discussion_r1038509208>
> :
>
> > @@ -0,0 +1,312 @@
>
> +_g_rank_cutoff = 1e-14  # cut off of eigenvalue to determine rank of span of two orthogonal basis
>
>
> These first seven lines contain some “magic numbers” that not only have
> names I do not understand, but even more it would be better if a user could
> set these. Can you put them into their corresponding functions as keyword
> arguments please?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#pullrequestreview-1203041376>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFJCWJV34JQSG6ZJVT3WLJMELANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
Locally (just like the tests) you can run the file `docs/make.jl` then (if you have all necessary packages loaded and it runs without errors) the rendered docs (HTML) are in `/docs/build` (just that on CI this folder gets pushed to our GitHub pages, since I activated that now here, even for your PR branch here – if that is successful the last entry in the checks is a link to said preview docs).

dnguyend commented:
Cool, thanks!

On Fri, Dec 2, 2022 at 3:41 PM Ronny Bergmann ***@***.***>
wrote:

> Locally (just like the tests) you can run the file docs/make.jl then (if
> you have all necessary packages loaded and it runs without errors) the
> rendered docs (HTML) are in /docs/build (just that on CI this folder gets
> pushed to our GitHub pages, since I activated that now here, even for your
> PR branch here – if that is successful the last entry in the checks is a
> link to said preview docs).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1335825144>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFPN6I6X2AJUMXJP5LTWLJNJJANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
Is there an easy way to make the documenter reflect the docstring after I
change it - it seems to cache the old version, it is very difficult to make
it to reflect the changes. I have to restart the REPL, then run dev ., then
run include("docs/make.jl") every time I want to see a change - and it has
been painful.

On Fri, Dec 2, 2022 at 3:43 PM Du Nguyen ***@***.***> wrote:

> Cool, thanks!
>
> On Fri, Dec 2, 2022 at 3:41 PM Ronny Bergmann ***@***.***>
> wrote:
>
>> Locally (just like the tests) you can run the file docs/make.jl then (if
>> you have all necessary packages loaded and it runs without errors) the
>> rendered docs (HTML) are in /docs/build (just that on CI this folder
>> gets pushed to our GitHub pages, since I activated that now here, even for
>> your PR branch here – if that is successful the last entry in the checks is
>> a link to said preview docs).
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1335825144>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEOOGFPN6I6X2AJUMXJP5LTWLJNJJANCNFSM6AAAAAASO7CVTA>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

kellertuer commented:
oh, `dev .`should not be necessary, if you have the package under version control, but you have to reload the package – I usually restart the REPL, but maybe Revise.jl works as well.

dnguyend commented:
Ok, thanks!

On Sat, Dec 3, 2022 at 11:37 AM Ronny Bergmann ***@***.***>
wrote:

> oh, dev .should not be necessary, if you have the package under version
> control, but you have to reload the package – I usually restart the REPL,
> but maybe Revise.jl works as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1336194914>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFKQVF4FGLRYMO2MQWDWLNZNXANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
I think I got the documentation looking reasonable (haven't pushed yet),
now I am trying to work on code coverage. Is there an easy way to do it
locally ? I see there is https://github.com/JuliaCI/Coverage.jl - Do you
know if it is a reasonable tool, or is there anything else I can try ?

I could send the code into CI but it will take hours to run through the
full project, I want to do it in small modules so I can get quicker
feedback and fix problems quickly - one by one.

On Sat, Dec 3, 2022 at 11:39 AM Du Nguyen ***@***.***> wrote:

> Ok, thanks!
>
> On Sat, Dec 3, 2022 at 11:37 AM Ronny Bergmann ***@***.***>
> wrote:
>
>> oh, dev .should not be necessary, if you have the package under version
>> control, but you have to reload the package – I usually restart the REPL,
>> but maybe Revise.jl works as well.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1336194914>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEOOGFKQVF4FGLRYMO2MQWDWLNZNXANCNFSM6AAAAAASO7CVTA>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

dnguyend commented:
I modified runtests.jl to contain only stiefel.jl then run
Pkg.test("Manifolds", coverage = true). Then I go through the output .cov
files - that is good for me for now, and I get feedback right away.

 But if there is a better way - I would love to learn.
Thx

On Sat, Dec 3, 2022 at 7:01 PM Du Nguyen ***@***.***> wrote:

> I think I got the documentation looking reasonable (haven't pushed yet),
> now I am trying to work on code coverage. Is there an easy way to do it
> locally ? I see there is https://github.com/JuliaCI/Coverage.jl - Do you
> know if it is a reasonable tool, or is there anything else I can try ?
>
> I could send the code into CI but it will take hours to run through the
> full project, I want to do it in small modules so I can get quicker
> feedback and fix problems quickly - one by one.
>
> On Sat, Dec 3, 2022 at 11:39 AM Du Nguyen ***@***.***> wrote:
>
>> Ok, thanks!
>>
>> On Sat, Dec 3, 2022 at 11:37 AM Ronny Bergmann ***@***.***>
>> wrote:
>>
>>> oh, dev .should not be necessary, if you have the package under version
>>> control, but you have to reload the package – I usually restart the REPL,
>>> but maybe Revise.jl works as well.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1336194914>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AEOOGFKQVF4FGLRYMO2MQWDWLNZNXANCNFSM6AAAAAASO7CVTA>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>
>>
>> --
>> Thanks,
>> Du
>>
>
>
> --
> Thanks,
> Du
>


-- 
Thanks,
Du

kellertuer commented:
instead of modifying runtest you can also just include that file yourself. We wrote the single manifold tests in a way that they also work standalone, that is only with `include`, for my workflow it would be great if you could comment in the points raised above when you have solved them (of course also when you have a question or would like to not approach something I asked about).

Code coverage for a little bit better, but we are not yet there.

dnguyend commented:
Thanks. I added replies to the issues you raised above.

I have been running test on single files. I would like to know the syntax to run one file and generate the coverage file, the .cov file generated next to the code showing what line of code is not covered, like these:

[ Info: CoverageTools.process_file: Detecting coverage for src/manifolds/StiefelEuclideanMetric.jl
[ Info: CoverageTools.process_cov: processing src/manifolds/StiefelEuclideanMetric.jl.26760.cov
[ Info: CoverageTools.process_file: Detecting coverage for src/manifolds/StiefelFrechet.jl
[ Info: CoverageTools.process_cov: processing src/manifolds/StiefelFrechet.jl.26760.cov
[ Info: CoverageTools.process_file: Detecting coverage for src/manifolds/StiefelSubmersionMetric.jl
[ Info: CoverageTools.process_cov: processing src/manifolds/StiefelSubmersionMetric.jl.26760.cov

Pkg.test("Manifolds", coverage=true) generates the .cov files, they are useful for me, but I had to modify runtests.jl to trick it to work - so it would be great to have a syntax to do it directly.

I am able to get coverage for expm_frechet.jl and StiefelFrechet.jl to near 98%, only the views are now flagged as uncovered. I don't know how to fix the views - presumably there is an ignore option in the setup file ?

[expm_frechet.jl](https://app.codecov.io/gh/JuliaManifolds/Manifolds.jl/blob/master/src/expm_frechet.jl)
[StiefelFrechet.jl](https://app.codecov.io/gh/JuliaManifolds/Manifolds.jl/blob/master/src/manifolds/StiefelFrechet.jl)

I haven't started on minimize_lbfgs.jl, this is a bit tricky as it is a general purpose optimizer - there are many scenarios - but I am working on it.

dnguyend commented:
Thanks - I will address the points you raised - By the way you mentioned
the docs for the PR run are activated - but I could not find the link -
could you please share ? The docs I generated locally look reasonable for
me.

> Just quickly, if its relevant ( will add to the doc), the magic numbers on
> expm_frechet comes from pade approximation (copied from scipy.linalg)
>
> Concerning this point – hm I am not much a fan of including code from
> other packages here, but if that is necessary, then relay only with a lot
> of documentation and notes where that is from (and a check of whether their
> License allows that of course)
>
I checked with the original article - and have added explanations - the
quoted paragraph was from Friday - which I believe have been addressed in
last night's run.

> Concerning the @views that are false positives and they will sadly stay
> (actually these false positives are the reason we are not at 100% Code
> Coverage
>
> minimize_lbfgs.jl, this is a bit tricky as it is a general purpose
> optimizer
>
> but did you write that yourself? Otherwise – and for code even more than
> for constants – I would be very unhappy to copy over code from other
> packages - and maybe if you write a more specialised solver it might both
> be faster and easier to test?
>
you could see the coverage is 80% for this case - the remaining cases are
rarer cases - which could happen, some cases are just status messages when
the optimization problem fails to converge. I just have to work to cover
them all.

I just want to be proactive and explain where I am in the process - not
mean as an excuse, I am sorry if it sounds that way.

-- 
Thanks,
Du

kellertuer commented:
No worries, I am aware that you are in the process, I am just trying in a few steps to give feedback early to reduce your work (hopefully at least ;) )

dnguyend commented:
The new push should bring the 3 files I work on much closer to target, but if there are still issues I will investigate.

I added 4 functions out in the docs, expm_frechet, minimize (from minimize_lbfgs), dot_exp and log_lbfgs. The helper functions have comments in docstring format but does not show on the docs. 
Thx

dnguyend commented:
I submitted a change this morning which should improve coverage for the 3
files to above 97%. Somehow the coverage summary seems to have crashed -
but the CI run is flagged as OK otherwise.

Please take a look when you get a chance - I hope it's not something the
code caused.

Other than that, I think the outstanding coverage issues are with @views,
and there is one rare condition in the optimizer which the next release
should fix.


On Sun, Dec 4, 2022 at 12:33 PM Ronny Bergmann ***@***.***>
wrote:

> No worries, I am aware that you are in the process, I am just trying in a
> few steps to give feedback early to reduce your work (hopefully at least ;)
> )
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1336474092>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFINWAQZ4S3ALDJ3MG3WLTIWPANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
The first thing I see is that your test runs issue quite some output. That should never be the case, especially not something like 
```
i=	27 n=	93 k=	15
```
besides that maybe one of the reports (we gather 2, one for Lie groups one for manifolds) came too late (before this PR the manifolds test case ran for about an hour, your last test run was more than 1.5 hours it seems which is really long).
So for now I am not 100% sure how to continue since it seems something in your tests causes the second report not to be reported or being reported too late.

dnguyend commented:
I will remove the print out and will resubmit. Hope this will fix the
problem.

The one change I made to improve coverage is there are some "if" cases that
are rarely hit, so I run 50 log_lbfgs randomly (with stating seed) and get
all of them,

Probably 10+ runs among those 50 fill the missing cases but to restrict to
those I have to save actual manifold/manifold points.

On Mon, Dec 5, 2022 at 11:28 AM Ronny Bergmann ***@***.***>
wrote:

> The first thing I see is that your test runs issue quite some output. That
> should never be the case, especially not something like
>
> i=	27 n=	93 k=	15
>
> besides that maybe one of the reports (we gather 2, one for Lie groups one
> for manifolds) came too late (before this PR the manifolds test case ran
> for about an hour, your last test run was more than 1.5 hours it seems
> which is really long).
> So for now I am not 100% sure how to continue since it seems something in
> your tests causes the second report not to be reported or being reported
> too late.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1337666048>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFLOXVJ6DHJPSDOORI3WLYJ4BANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

kellertuer commented:
Please do not do random tests – that is absolutely horrible to debug _if_ one of the random cases fails, because it is super hard to debug – for any maintainer. I have such a case for about a year now I another package and was not yet able to reproduce the case in a non-random setting to finally fix it.

So: Please no random stuff in the tests.

kellertuer commented on src/expm_frechet.jl:
```diff
@@ -0,0 +1,518 @@
+    if isnothing(s)
+        # scaling
+        s = max(0, Int(ceil(log2(A_norm_1 / ell_table_61[14]))))
+        # pade order 13
```
What is the goal of this comment if the next line calls a function of same name? And why are there so many other diff_pade13 functions? To more this file is a little chaotic and most of its content a little strange – sorry, but it is really mainly the problem that I will have to be able to maintain that and in its current form, this is just not possible.

dnguyend replied:
will clean up - and make diff_pade13 calling diff_pade13!

kellertuer commented on src/manifolds/StiefelFrechet.jl:
```diff
@@ -0,0 +1,314 @@
+    t,
+) where {n,k}
+    q = similar(p)
+    dot_q = similar(p)
```
I think `dot_q` is a tangent vector, right? First of all, then `dot_q`should be called `diff_exp` or `differential_exp` but even more, `dot_q` should be initialised as a tangent vector (`zero_vector(M,p)` probably) and better be called `X` or `Y` following our naming convention.

dnguyend replied:
OK will do

dnguyend commented:
Thanks - it is kind of unexpected - thanks for the suggestion. I will have to think about working on a new package - that will likely to take more time commitment than I can afford.

By the way, I would like to share a paper [Operator-valued formulas for Riemannian Gradient and Hessian and families of tractable ]( https://arxiv.org/abs/2009.10159) that shows how to compute the Levi-Civita connection for most metrics, including Stiefel/Grassmann/Flag manifolds, semi-definite matrix manifold with a complete metric. I believe it also gives the connection for the Symplectic Stiefel manifold that is in your package.
Best,
Du

kellertuer commented:
Thanks for the link.

I am sorry that that will take more time, but keep in mind I am doing this in mz free time and have spent so much time on this already, and barely looked at about 20-30 lines of code. In that speed this PR is done (given that I spent an hour a day or such, like I did the last week here) this will take a very very very long time with a lot of time from my side. And sorry, I can not do that besides lectures, master and phd students and my own research.

Besides that there are the questions with licensing that I am still not sure about since you did not answer that completely yet. In your own package – I would not care about your licensing – in mine I do.

dnguyend commented:
Thank you for the offer - too bad this didn't work out.

On Mon, Dec 12, 2022 at 2:43 AM Ronny Bergmann ***@***.***>
wrote:

> Closed #555 <https://github.com/JuliaManifolds/Manifolds.jl/pull/555>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#event-8011144049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFNDU2AP3LHSCS4OZ53WM3JSVANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du

dnguyend commented:
I understand. Thanks!

On Mon, Dec 12, 2022 at 12:11 PM Ronny Bergmann ***@***.***>
wrote:

> As I mentioned – I can currently not provide the time to help you finish
> this, especially
>
>    - documentation
>    - testing
>    - review all code lines and the design of the functions.
>    - Check dependencies and especially Licensing of code you include
>
> and there seems to be no one else as well. If you at some point have these
> criteria fulfilled – feel free to reopen this PR or do a new one. But I
> feel starting from a separate package is really the best way to go here. An
> we are happy to link your project then.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaManifolds/Manifolds.jl/pull/555#issuecomment-1346900558>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOOGFI2OG426KBMWAUJS43WM5MELANCNFSM6AAAAAASO7CVTA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks,
Du
